File: valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        matchingChar = {")": "(", "}": "{", "]": "["}

        for c in s:
            if c in matchingChar:
                if stack and stack[-1] == matchingChar[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)

        return True if not stack else False

File: validParentheses.js
class Solution {
  /**
   * @param {string} s
   * @return {boolean}
   */
  isValid(s) {
    let stack = [];
    let charSet = { ")": "(", "}": "{", "]": "[" };

    for (const c of s) {
      if (c in charSet) {
        if (stack.length && stack[stack.length - 1] === charSet[c]) {
          stack.pop();
        } else {
          return false;
        }
      } else {
        stack.push(c);
      }
    }
    return stack.length === 0;
  }
}
